notebooks/work-in-progress/ferc1-transform-refactor/ferc1_transform_refactor.py

```diff
@@ -0,0 +1,429 @@
+
+
+categorize_strings = multicol_transform_fn_factory(categorize_strings_col, drop=True)
+
+
+################################################################################
+# TableTransformer classes
+################################################################################
+class AbstractTableTransformer(ABC):
+    """An exmaple abstract base transformer class."""
+
+    table_id: str
+
+    @property
+    def params(self) -> TableTransformParams:
+        """Obtain table transform parameters based on the table ID."""
+        return TableTransformParams.from_id(table_id=self.table_id)
+
+    ################################################################################
+    # Abstract methods that must be defined by subclasses
+    @abstractmethod
+    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Apply all specified transformations to the appropriate input dataframes."""
+        ...
+
+    ################################################################################
+    # Default method implementations which can be used or overridden by subclasses
+    def rename_columns(self, df: pd.DataFrame, params: RenameColumns) -> pd.DataFrame:
+        """Rename the whole collection of dataframe columns using input params.
+
+        Log if there's any mismatch between the columns in the dataframe, and the
+        columns that have been defined in the mapping for renaming.
+
+        """
+        return df.rename(columns=params.columns)
+
+
+class TestTableTransformerOne(AbstractTableTransformer):
+    """A concrete table transformer class for testing."""
+
+    table_id: str = "test_table_one"
+
+    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Apply all the Test Transforms."""
+        return (
+            df.pipe(self.rename_columns, self.params.rename_columns)
+            .pipe(categorize_strings, self.params.categorize_strings)
+            .pipe(convert_units, self.params.convert_units)
+            .pipe(self.bespoke_transform)
+        )
+
+    def bespoke_transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """A transformation that's specific to this table."""
+        return df.replace(to_replace="coal", value="💩")
+
+
+class TestTableTransformerTwo(AbstractTableTransformer):
+    """A concrete table transformer class for testing."""
+
+    table_id: str = "test_table_two"
+
+    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Apply all the Test Transforms."""
+        return (
+            df.pipe(self.rename_columns, self.params.rename_columns)
+            .pipe(categorize_strings, self.params.categorize_strings)
+            .pipe(convert_units, self.params.convert_units)
+            .pipe(self.bespoke_transform)
+        )
+
+    def bespoke_transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """A transformation that's specific to this table."""
+        return df.replace(to_replace="coal", value="👿")
+
+
+def table_transformer_fn_factory(
```

Comment:
Maybe! My brain starts to hurt when I think about decorators too much. I think this will work as is:

```python
@job
def etl():
       raw_df = extract()
       transformed_df = table_transformer_fn_factory("table_1")(raw_df)
       load(transformed_df)
```